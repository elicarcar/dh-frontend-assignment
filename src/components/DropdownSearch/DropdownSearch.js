import React, { useState } from "react";
import Wrapper from "./DropdownSearchWrapper/Wrapper.js";
import Nav from "./DropdownSearchNav/Nav.js";
import List from "./DropdownSearchList/List.js";
import ListItems from "./DropdownListItems/ListItems.js";
import Label from "./DropdownLabel/DropdownLabel.js";
import Input from "./DropdownInput/Input.js";

const DropdownSearch = ({ setName, contactNames = [], inputRef, navRef }) => {
  const [nameToFilter, setNameToFilter] = useState(/(?:)/);
  const [displaySearchIcon, setDisplaySearchIcon] = useState(false);
  const [contactName, setContactName] = useState("");
  const noMatch =
    contactNames.filter(name => nameToFilter.test(name.name)).length === 0;

  const handleInputChange = e => {
    const searchedName = e.target.value;
    setNameToFilter(new RegExp(e.target.value, "i"));
    setContactName(searchedName);
  };

  const handleFocus = () => {
    return setTimeout(function() {
      setDisplaySearchIcon(true);
    }, 100);
  };

  const handleBlur = () => {
    setDisplaySearchIcon(false);
  };

  const getContact = name => {
    setContactName(name);
  };

  const getSelectedName = name => {
    setName(name);
  };

  return (
    <Wrapper displaySearchIcon={displaySearchIcon} divRef={divRef}>
      <Input
        data-testid="contact-input"
        value={contactName}
        handleInputChange={e => handleInputChange(e)}
        handleBlur={handleBlur}
        handleFocus={handleFocus}
        inputRef={inputRef}
      />
      <Label htmlFor="contact">Contact</Label>
      <Nav ref={navRef} data-testid="navBar">
        <List>
          {noMatch ? (
            <ListItems key="no-match" noMatch data-testid="no-match">
              No match
            </ListItems>
          ) : (
            contactNames
              .sort((a, b) => (a.name < b.name ? -1 : 0))
              .filter(
                name => nameToFilter.test(name.name) && name.name !== null
              )
              .map((person, i) => (
                <ListItems
                  key={i}
                  handleClick={() => getContact(person.name)}
                  handleSelect={() => getSelectedName(person.name)}
                >
                  {person.name}
                </ListItems>
              ))
          )}
        </List>
      </Nav>
    </Wrapper>
  );
};

export default DropdownSearch;
