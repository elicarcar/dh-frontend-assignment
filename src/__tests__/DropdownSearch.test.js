import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, act, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { toBeVisible } from "@testing-library/jest-dom/dist/to-be-visible";
import "jest-styled-components";
import DropdownSearch from "../components/DropdownSearch/DropdownSearch.js";
expect.extend({ toBeVisible });

afterEach(cleanup);
describe("<DropdownSearch/>", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<DropdownSearch />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays navbar when focusing on input", () => {
    const { getByTestId } = render(<DropdownSearch />);
    const input = getByTestId("contact-input");
    act(() => {
      input.focus();
    });
    expect(document.querySelector('[data-testid="navBar"]')).toHaveStyleRule(
      "visibility",
      "visible"
    );
  });

  it("changes the value", () => {
    const { getByPlaceholderText } = render(<DropdownSearch />);
    const input = getByPlaceholderText("Type or Search..");
    act(() => {
      fireEvent.change(input, { target: { value: "hello" } });
    });
    expect(input).toHaveValue("hello");
  });

  it("hides navbar when input blurs", () => {
    const { getByTestId } = render(<DropdownSearch />);

    const nav = getByTestId("navBar");
    const input = getByTestId("contact-input");

    act(() => {
      input.blur();
    });

    expect(nav).not.toBeVisible();
  });
});
