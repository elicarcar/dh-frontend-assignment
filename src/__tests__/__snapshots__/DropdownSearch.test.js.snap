// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownSearch/> renders correctly 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  margin: 20px;
}

.c0:before {
  display: none;
  position: absolute;
  content: " ⚲ ";
  -webkit-transform: scale(0.5);
  -ms-transform: scale(0.5);
  transform: scale(0.5);
  top: 4px;
  left: 2px;
  height: 10px;
  width: 24px;
  margin-right: 5px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

.c0:after {
  content: "▼";
  display: block;
  position: absolute;
  top: 11px;
  right: 45px;
}

.c5 {
  position: relative;
  visibility: visible;
  opacity: 1;
  -webkit-transition: visibility 0s,opacity 0.5s;
  transition: visibility 0s,opacity 0.5s;
  -webkit-transition-delay: 0.1s;
  transition-delay: 0.1s;
}

.c7 {
  list-style: none;
  margin: 2px;
  padding: 0;
  overflow: hidden;
  overflow-y: scroll;
  border-radius: 3px;
  border: 1px solid #bfc5cd;
  box-shadow: 0 5px 15px 0 rgba(74,74,74,0.15);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  min-width: 230px;
  max-width: 350px;
  height: 200px;
  min-height: 100px;
  max-height: 250px;
}

.c8 {
  font-family: "Helvetica";
  color: #798697;
  line-height: 30px;
  padding: 1px;
  padding-left: 11px;
  font-size: 14px;
  cursor: not-allowed;
}

.c8:hover {
  color: #4a4a4a;
  background: #f7f7f7;
}

.c3 {
  position: absolute;
  padding-left: 12px;
  margin-bottom: 15px;
  top: 9px;
  left: 0px;
  -webkit-transition: -webkit-transform 0.1s;
  -webkit-transition: transform 0.1s;
  transition: transform 0.1s;
  -webkit-transition-timing-function: cubic-bezier(0.25,0.1,0.25,0.1);
  transition-timing-function: cubic-bezier(0.25,0.1,0.25,0.1);
  opacity: 0.5;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  border: 1px solid #bfc5cd;
  border-radius: 7px;
  color: #4a4a4a;
  padding: 10px;
  outline: none;
  -webkit-transition: all 0.6s;
  transition: all 0.6s;
  -webkit-transition-delay: 0.2s;
  transition-delay: 0.2s;
}

.c1::-webkit-input-placeholder {
  opacity: 0;
}

.c1:hover {
  border: 1px solid #4a4a4a;
}

.c1:focus {
  text-indent: 12px;
}

.c1:focus::-webkit-input-placeholder {
  opacity: 1;
}

.c1:not(:focus) ~ .c4 {
  visibility: hidden;
  opacity: 0;
}

.c1:focus ~ .c2,
.c1:not(:placeholder-shown) ~ .c2 {
  -webkit-transform: scale(0.9) translateY(-135%) translateX(-10px);
  -ms-transform: scale(0.9) translateY(-135%) translateX(-10px);
  transform: scale(0.9) translateY(-135%) translateX(-10px);
  opacity: 1;
  padding-bottom: 2px;
}

<div
  className="c0"
  id="wrapper-div"
>
  <input
    autoComplete="off"
    className="c1"
    data-testid="contact-input"
    id="contact-input"
    name="contact"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type or Search.."
    type="text"
    value=""
  />
  <label
    className="c2 c3"
    htmlFor="contact"
  >
    Contact
  </label>
  <nav
    className="c4 c5"
    data-testid="navBar"
  >
    <div
      className="c6"
    >
      <ul
        className="c7"
      >
        <li
          className="c8"
          data-testid="no-match"
        >
          No match
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
